File: code/dart/modular_api_rest_client/test/http_client_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:modular_api_rest_client/modular_api_rest_client.dart';
import 'package:test/test.dart';

void main() {
  group('httpClient', () {
    test('sends a GET request, decodes JSON, and preserves metadata', () async {
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      addTearDown(server.close);

      server.listen((request) async {
        expect(request.method, 'GET');
        expect(request.uri.path, '/users');
        expect(request.uri.queryParameters['name'], 'ana');
        expect(request.headers.value('x-default'), 'package');
        expect(request.headers.value('x-request'), 'test');

        request.response.statusCode = HttpStatus.ok;
        request.response.headers.contentType = ContentType.json;
        request.response.headers.set('x-request-id', 'req-123');
        request.response.write(jsonEncode({'ok': true, 'name': 'Ana'}));
        await request.response.close();
      });

      final config = ServiceClientConfig(
        serviceId: 'users',
        baseUrl: Uri.parse('http://127.0.0.1:${server.port}'),
        redactedSummary: 'users@local',
        defaultHeaders: const {'x-default': 'package'},
      );

      final request = ServiceRequest(
        operationId: 'get-users',
        method: 'GET',
        path: '/users',
        query: const {'name': 'ana'},
        headers: const {'x-request': 'test'},
      );

      final result = await httpClient<Map<String, Object?>>(
        config: config,
        request: request,
        decoder: (json) => Map<String, Object?>.from(json as Map),
      );

      expect(result.isSuccess, isTrue);
      expect(result.value.data['ok'], isTrue);
      expect(result.value.data['name'], 'Ana');
      expect(result.value.metadata.statusCode, HttpStatus.ok);
      expect(result.value.metadata.transportId, 'http');
      expect(result.value.metadata.requestId, 'req-123');
      expect(result.value.metadata.headers['x-request-id'], 'req-123');
    });

    test('returns a decode failure for invalid JSON responses', () async {
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      addTearDown(server.close);

      server.listen((request) async {
        request.response.statusCode = HttpStatus.ok;
        request.response.headers.contentType = ContentType.json;
        request.response.write('{broken-json');
        await request.response.close();
      });

      final result = await httpClient<Object?>(
        config: ServiceClientConfig(
          serviceId: 'broken',
          baseUrl: Uri.parse('http://127.0.0.1:${server.port}'),
          redactedSummary: 'broken@local',
        ),
        request: const ServiceRequest(
          operationId: 'decode-failure',
          method: 'GET',
          path: '/broken',
        ),
      );

      expect(result.isFailure, isTrue);
      expect(result.failure.category, ServiceFailureCategory.decode);
      expect(result.failure.code, 'invalid_json');
    });

    test('injects auth headers from the auth provider', () async {
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      addTearDown(server.close);

      server.listen((request) async {
        expect(request.headers.value('authorization'), 'Bearer token-123');

        request.response.statusCode = HttpStatus.ok;
        request.response.headers.contentType = ContentType.json;
        request.response.write(jsonEncode({'ok': true}));
        await request.response.close();
      });

      final result = await httpClient<Map<String, Object?>>(
        config: ServiceClientConfig(
          serviceId: 'users',
          baseUrl: Uri.parse('http://127.0.0.1:${server.port}'),
          redactedSummary: 'users@local',
          authProvider: (operation) async {
            expect(operation.operationId, 'auth-check');
            return const {'authorization': 'Bearer token-123'};
          },
        ),
        request: const ServiceRequest(
          operationId: 'auth-check',
          method: 'GET',
          path: '/users',
        ),
        decoder: (json) => Map<String, Object?>.from(json as Map),
      );

      expect(result.isSuccess, isTrue);
      expect(result.value.data['ok'], isTrue);
    });

    test('returns a timeout failure when the request exceeds the configured timeout', () async {
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      addTearDown(server.close);

      server.listen((request) async {
        await Future<void>.delayed(const Duration(milliseconds: 200));
        request.response.statusCode = HttpStatus.ok;
        request.response.headers.contentType = ContentType.json;
        request.response.write(jsonEncode({'late': true}));
        await request.response.close();
      });

      final result = await httpClient<Object?>(
        config: ServiceClientConfig(
          serviceId: 'slow',
          baseUrl: Uri.parse('http://127.0.0.1:${server.port}'),
          redactedSummary: 'slow@local',
          timeout: const Duration(milliseconds: 20),
        ),
        request: const ServiceRequest(
          operationId: 'timeout-check',
          method: 'GET',
          path: '/slow',
        ),
      );

      expect(result.isFailure, isTrue);
      expect(result.failure.category, ServiceFailureCategory.timeout);
      expect(result.failure.code, 'timeout');
      expect(result.failure.retryable, isTrue);
    });

    test('normalizes auth failures for non-2xx HTTP responses', () async {
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      addTearDown(server.close);

      server.listen((request) async {
        request.response.statusCode = HttpStatus.unauthorized;
        request.response.write('missing token');
        await request.response.close();
      });

      final result = await httpClient<Object?>(
        config: ServiceClientConfig(
          serviceId: 'auth',
          baseUrl: Uri.parse('http://127.0.0.1:${server.port}'),
          redactedSummary: 'auth@local',
        ),
        request: const ServiceRequest(
          operationId: 'unauthorized',
          method: 'GET',
          path: '/auth',
        ),
      );

      expect(result.isFailure, isTrue);
      expect(result.failure.category, ServiceFailureCategory.auth);
      expect(result.failure.code, 'unauthorized');
      expect(result.failure.statusCode, HttpStatus.unauthorized);
      expect(result.failure.details, 'missing token');
    });

    test('normalizes rate-limit failures for non-2xx HTTP responses', () async {
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      addTearDown(server.close);

      server.listen((request) async {
        request.response.statusCode = HttpStatus.tooManyRequests;
        request.response.write('retry later');
        await request.response.close();
      });

      final result = await httpClient<Object?>(
        config: ServiceClientConfig(
          serviceId: 'rate-limit',
          baseUrl: Uri.parse('http://127.0.0.1:${server.port}'),
          redactedSummary: 'rate-limit@local',
        ),
        request: const ServiceRequest(
          operationId: 'too-many',
          method: 'GET',
          path: '/rate-limit',
        ),
      );

      expect(result.isFailure, isTrue);
      expect(result.failure.category, ServiceFailureCategory.rateLimit);
      expect(result.failure.code, 'rate_limit');
      expect(result.failure.retryable, isTrue);
      expect(result.failure.statusCode, HttpStatus.tooManyRequests);
    });
  });

  group('HttpServiceClient', () {
    test('describes its config and closes cleanly', () async {
      final client = HttpServiceClient(
        ServiceClientConfig(
          serviceId: 'users',
          baseUrl: Uri.parse('https://example.test'),
          redactedSummary: 'users@example',
        ),
      );

      expect(client.describe().serviceId, 'users');
      expect(client.describe().transportId, 'http');

      final closed = await client.close();
      expect(closed.isSuccess, isTrue);
    });
  });
}

File: code/ts/modular_api_rest_client/test/httpClient.test.ts
import { createServer, type IncomingMessage, type ServerResponse } from 'node:http';
import { AddressInfo } from 'node:net';

import {
  HttpServiceClient,
  ServiceClientConfig,
  ServiceFailureCategory,
  ServiceRequest,
  httpClient,
} from '../src';

import { afterEach, describe, expect, it } from 'vitest';

describe('httpClient', () => {
  const servers: Array<ReturnType<typeof createServer>> = [];

  afterEach(async () => {
    await Promise.all(
      servers.map(
        (server) =>
          new Promise<void>((resolve, reject) => {
            server.close((error) => (error ? reject(error) : resolve()));
          }),
      ),
    );
    servers.length = 0;
  });

  it('sends a GET request, decodes JSON, and preserves metadata', async () => {
    const server = await startServer(async (req, res) => {
      const url = new URL(req.url ?? '/', 'http://127.0.0.1');

      expect(req.method).toBe('GET');
      expect(url.pathname).toBe('/users');
      expect(url.searchParams.get('name')).toBe('ana');
      expect(req.headers['x-default']).toBe('package');
      expect(req.headers['x-request']).toBe('test');

      res.statusCode = 200;
      res.setHeader('content-type', 'application/json');
      res.setHeader('x-request-id', 'req-123');
      res.end(JSON.stringify({ ok: true, name: 'Ana' }));
    }, servers);

    const result = await httpClient({
      config: new ServiceClientConfig({
        serviceId: 'users',
        baseUrl: server.baseUrl,
        redactedSummary: 'users@local',
        defaultHeaders: { 'x-default': 'package' },
      }),
      request: new ServiceRequest({
        operationId: 'get-users',
        method: 'GET',
        path: '/users',
        query: { name: 'ana' },
        headers: { 'x-request': 'test' },
      }),
      decoder: (value) => value as Record<string, unknown>,
    });

    expect(result.isSuccess).toBe(true);
    expect(result.value.data.ok).toBe(true);
    expect(result.value.data.name).toBe('Ana');
    expect(result.value.metadata.statusCode).toBe(200);
    expect(result.value.metadata.transportId).toBe('http');
    expect(result.value.metadata.requestId).toBe('req-123');
    expect(result.value.metadata.headers['x-request-id']).toBe('req-123');
  });

  it('returns a decode failure for invalid JSON responses', async () => {
    const server = await startServer(async (_req, res) => {
      res.statusCode = 200;
      res.setHeader('content-type', 'application/json');
      res.end('{broken-json');
    }, servers);

    const result = await httpClient({
      config: new ServiceClientConfig({
        serviceId: 'broken',
        baseUrl: server.baseUrl,
        redactedSummary: 'broken@local',
      }),
      request: new ServiceRequest({
        operationId: 'decode-failure',
        method: 'GET',
        path: '/broken',
      }),
    });

    expect(result.isFailure).toBe(true);
    expect(result.failure.category).toBe(ServiceFailureCategory.decode);
    expect(result.failure.code).toBe('invalid_json');
  });

  it('injects auth headers from the auth provider', async () => {
    const server = await startServer(async (req, res) => {
      expect(req.headers.authorization).toBe('Bearer token-123');

      res.statusCode = 200;
      res.setHeader('content-type', 'application/json');
      res.end(JSON.stringify({ ok: true }));
    }, servers);

    const result = await httpClient({
      config: new ServiceClientConfig({
        serviceId: 'users',
        baseUrl: server.baseUrl,
        redactedSummary: 'users@local',
        authProvider: async (operation) => {
          expect(operation.operationId).toBe('auth-check');
          return { authorization: 'Bearer token-123' };
        },
      }),
      request: new ServiceRequest({
        operationId: 'auth-check',
        method: 'GET',
        path: '/users',
      }),
      decoder: (value) => value as Record<string, unknown>,
    });

    expect(result.isSuccess).toBe(true);
    expect(result.value.data.ok).toBe(true);
  });

  it('returns a timeout failure when the request exceeds the configured timeout', async () => {
    const server = await startServer(async (_req, res) => {
      await new Promise((resolve) => setTimeout(resolve, 200));
      res.statusCode = 200;
      res.setHeader('content-type', 'application/json');
      res.end(JSON.stringify({ late: true }));
    }, servers);

    const result = await httpClient({
      config: new ServiceClientConfig({
        serviceId: 'slow',
        baseUrl: server.baseUrl,
        redactedSummary: 'slow@local',
        timeout: 20,
      }),
      request: new ServiceRequest({
        operationId: 'timeout-check',
        method: 'GET',
        path: '/slow',
      }),
    });

    expect(result.isFailure).toBe(true);
    expect(result.failure.category).toBe(ServiceFailureCategory.timeout);
    expect(result.failure.code).toBe('timeout');
    expect(result.failure.retryable).toBe(true);
  });

  it('normalizes auth failures for non-2xx HTTP responses', async () => {
    const server = await startServer(async (_req, res) => {
      res.statusCode = 401;
      res.end('missing token');
    }, servers);

    const result = await httpClient({
      config: new ServiceClientConfig({
        serviceId: 'auth',
        baseUrl: server.baseUrl,
        redactedSummary: 'auth@local',
      }),
      request: new ServiceRequest({
        operationId: 'unauthorized',
        method: 'GET',
        path: '/auth',
      }),
    });

    expect(result.isFailure).toBe(true);
    expect(result.failure.category).toBe(ServiceFailureCategory.auth);
    expect(result.failure.code).toBe('unauthorized');
    expect(result.failure.statusCode).toBe(401);
    expect(result.failure.details).toBe('missing token');
  });

  it('normalizes rate-limit failures for non-2xx HTTP responses', async () => {
    const server = await startServer(async (_req, res) => {
      res.statusCode = 429;
      res.end('retry later');
    }, servers);

    const result = await httpClient({
      config: new ServiceClientConfig({
        serviceId: 'rate-limit',
        baseUrl: server.baseUrl,
        redactedSummary: 'rate-limit@local',
      }),
      request: new ServiceRequest({
        operationId: 'too-many',
        method: 'GET',
        path: '/rate-limit',
      }),
    });

    expect(result.isFailure).toBe(true);
    expect(result.failure.category).toBe(ServiceFailureCategory.rateLimit);
    expect(result.failure.code).toBe('rate_limit');
    expect(result.failure.retryable).toBe(true);
    expect(result.failure.statusCode).toBe(429);
  });
});

describe('HttpServiceClient', () => {
  it('describes its config and closes cleanly', async () => {
    const client = new HttpServiceClient(
      new ServiceClientConfig({
        serviceId: 'users',
        baseUrl: 'https://example.test',
        redactedSummary: 'users@example',
      }),
    );

    expect(client.describe().serviceId).toBe('users');
    expect(client.describe().transportId).toBe('http');

    const closed = await client.close();
    expect(closed.isSuccess).toBe(true);
  });
});

async function startServer(
  handler: (request: IncomingMessage, response: ServerResponse) => void | Promise<void>,
  registry: Array<ReturnType<typeof createServer>>,
): Promise<{ baseUrl: string }> {
  const server = createServer((request, response) => {
    void Promise.resolve(handler(request, response)).catch((error: unknown) => {
      response.statusCode = 500;
      response.end(String(error));
    });
  });
  registry.push(server);

  await new Promise<void>((resolve, reject) => {
    server.listen(0, '127.0.0.1', (error?: Error) => {
      if (error) {
        reject(error);
        return;
      }
      resolve();
    });
  });

  const address = server.address() as AddressInfo;
  return { baseUrl: `http://127.0.0.1:${address.port}` };
}